File: backend-app/src/main/java/com/backend_app/backend_app/service/ServicioService.java
package com.backend_app.backend_app.service;

import com.backend_app.backend_app.domain.Servicio;
import com.backend_app.backend_app.dto.ServicioDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface ServicioService {

    List<ServicioDTO> listadoServicios();

    Optional<ServicioDTO> buscarServicioPorId(long idServicio);

    ServicioDTO convertirADTO(Servicio servicio);

    void insertarServicio(ServicioDTO servicioDTO);

    void editarServicio(ServicioDTO servicioDTO);

    void eliminarServicio(ServicioDTO servicioDTO);

    Page<ServicioDTO> findAll(Pageable pageable);

    List<ServicioDTO> obtenerServiciosAsociados();

    List<ServicioDTO> buscarServicioPorNombre(String nombreServicio);

    List<ServicioDTO> listadoServiciosCompletos();

}

File: frontend-app/src/app/services/servicio.service.ts
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { Servicio } from "../domain/servicio";
import { HttpClient, HttpParams } from "@angular/common/http";


@Injectable({
    providedIn: 'root'
})
export class ServicioService{

    private url: string = 'http://localhost:8080/api/servicios';

    constructor(private http: HttpClient){

    }

    listadoServicios(): Observable<Servicio[]>{
        return this.http.get<Servicio[]>(this.url);
    }

    buscarServicioPorId(id: number): Observable<Servicio>{
        return this.http.get<Servicio>(`${this.url}/${id}`);
    }
    
    guardarServicio(servicio: Servicio){
        return this.http.post('http://localhost:8080/api/servicios/guardar', servicio);
    }
            
    editarServicio(servicio: Servicio) {
        return this.http.put(`${this.url}/actualizar`, servicio);
    }
            
    eliminarServicio(idServicio: number){
        return this.http.delete(`${this.url}/${idServicio}`);
    }

    listadoServicioPaginacion(page: number): Observable<any>{
        return this.http.get<any[]>(`${this.url}/page/${page}`);
    }   

    listadoServiciosAsociados(): Observable<Servicio[]>{
        return this.http.get<Servicio[]>(`${this.url}/servicios-asociados`)
    }

    buscarServicioPorNombre(nombreServicio: string): Observable<Servicio[]>{
        const params = new HttpParams().set('nombreServicio', nombreServicio);
                    
        return this.http.get<Servicio[]>(`${this.url}/buscar`, { params });
    }

    listadoServiciosCompletos(): Observable<Servicio[]>{
        return this.http.get<Servicio[]>(`${this.url}/completos`);
    }

}